src/neuralnet/project.cpp

```diff
@@ -0,0 +1,210 @@
+#include "project.h"
+
+#include <algorithm>
+#include <atomic>
+#include <boost/algorithm/string/case_conv.hpp>
+
+using namespace NN;
+
+//!
+//! \brief A mutable representation of a \c Project.
+//!
+//! The internal implementation uses this class to sort containers of \c Project
+//! instances. \c std::sort() cannot sort a collection of \c const objects.
+//!
+struct MutableProject
+{
+    //!
+    //! \brief Create a mutable representation of the provided \c Project.
+    //!
+    //! \param project The \c const instance to copy.
+    //!
+    MutableProject(const Project& project)
+        : m_name(project.m_name),
+          m_url(project.m_url),
+          m_timestamp(project.m_timestamp)
+    {
+    }
+
+    std::string m_name;   //!< As it exists in the contract key field.
+    std::string m_url;    //!< As it exists in the contract value field.
+    int64_t m_timestamp;  //!< Timestamp of the contract.
+
+    //!
+    //! \brief Get the lowercase equivalent of the project name.
+    //!
+    //! Lazy case-conversion. Used for sorting.
+    //!
+    //! \return The lowercase project name.
+    //!
+    std::string& LowercaseName() const
+    {
+        if (m_lower_name.empty()) {
+            m_lower_name = m_name;
+            boost::to_lower(m_lower_name);
+        }
+
+        return m_lower_name;
+    }
+
+private:
+    mutable std::string m_lower_name; //!< Caches the lowercase project name.
+};
+
+// -----------------------------------------------------------------------------
+// Class: Project
+// -----------------------------------------------------------------------------
+
+Project::Project(const std::string name, const std::string url, const int64_t ts)
+    : m_name(name), m_url(url), m_timestamp(ts)
+{
+}
+
+std::string Project::DisplayName() const
+{
+    std::string display_name = m_name;
+    std::replace(display_name.begin(), display_name.end(), '_', ' ');
+
+    return display_name;
+}
+
+std::string Project::BaseUrl() const
+{
+    // Remove the "@" from the URL in the contract. We assume that it always
+    // occurs at the very end:
+    return m_url.substr(0, m_url.size() - 1);
+}
+
+std::string Project::DisplayUrl() const
+{
+    // TODO: remove this after project contracts support arbitrary URLs.
+    // WCG project URL refers to a location inaccessible to the end user.
+    if (m_name == "World_Community_Grid") {
+        return "https://www.worldcommunitygrid.org/";
+    }
+
+    return BaseUrl();
+}
+
+std::string Project::StatsUrl(const std::string& type) const
+{
+    if (type.empty()) {
+        return BaseUrl() + "stats/";
+    }
+
+    return BaseUrl() + "stats/" + type + ".gz";
```

Comment:
This is wrong assumption. Project stats urls are not required to adhere to .../stats/type.gz .

Reply:
Agreed--until we change the format of URLs in the project contracts, I'm just doing it similarly to the way the old code does.